### PR TITLE
Update pico_sleep

### PR DESCRIPTION
**Pull request by peterharperuk**
Fixed to work for RP2040 and RP2350 using pico_aon_timer.

### CODE REVIEW

File: src/rp2_common/pico_sleep/include/pico/sleep.h [hunk 1]
```diff
@@ -62,10 +71,31 @@ static inline void sleep_run_from_rosc(void) {
  *
  * One of the sleep_run_* functions must be called prior to this call
  *
- * \param t The time to wake up
+ * \param ts The time to wake up
  * \param callback Function to call on wakeup.
  */
-void sleep_goto_sleep_until(datetime_t *t, rtc_callback_t callback);
+void sleep_goto_aon_sleep_until(struct timespec *ts, aon_timer_alarm_handler_t callback);
```
Comment (peterharperuk) [inline src/rp2_common/pico_sleep/include/pico/sleep.h:77]: remove aon from the name

File: src/rp2_common/pico_sleep/include/pico/sleep.h [hunk 2]
```diff
@@ -62,10 +71,31 @@ static inline void sleep_run_from_rosc(void) {
...
+
+/*! \brief Send system to dormant until the specified time, note for RP2040 the RTC must be driven by an external clock
+ *  \ingroup hardware_sleep
+ *
+ * One of the sleep_run_* functions must be called prior to this call
+ *
+ * \param ts The time to wake up
+ * \param callback Function to call on wakeup.
+ */
+void sleep_goto_aon_dormant_until(struct timespec *ts, aon_timer_alarm_handler_t callback);
```
Comment (peterharperuk) [inline src/rp2_common/pico_sleep/include/pico/sleep.h:98]: ditto

File: src/rp2_common/hardware_rosc/rosc.c [hunk 3]
```diff
@@ -58,4 +58,12 @@ void rosc_set_dormant(void) {
     rosc_write(&rosc_hw->dormant, ROSC_DORMANT_VALUE_DORMANT);
     // Wait for it to become stable once woken up
     while(!(rosc_hw->status & ROSC_STATUS_STABLE_BITS));
-}
\ No newline at end of file
+}
+
+void rosc_restart(void) {
```
Comment (lurch) [inline src/rp2_common/hardware_rosc/rosc.c:63]: Is it specifically a "restart", or could this function be named `rosc_enable` in order to mirror `rosc_disable` ?

File: src/rp2_common/pico_sleep/include/pico/sleep.h [hunk 4]
```diff
@@ -50,6 +53,12 @@ static inline void sleep_run_from_xosc(void) {
     sleep_run_from_dormant_source(DORMANT_SOURCE_XOSC);
 }
 
+#if !PICO_RP2040
+static inline void sleep_run_from_lpsc(void) {
```
Comment (lurch) [inline src/rp2_common/pico_sleep/include/pico/sleep.h:57]: Is the `_lpsc` here supposed to be `_lposc` ?

Comment (peterharperuk) [reply src/rp2_common/pico_sleep/include/pico/sleep.h:57]: yes

File: src/rp2_common/pico_sleep/include/pico/sleep.h [hunk 5]
```diff
@@ -62,10 +71,31 @@ static inline void sleep_run_from_rosc(void) {
...
+/*! \brief Send system to sleep for a specified duration in milliseconds. This provides an alternative to sleep_goto_sleep_until
+to allow for shorter duration sleeps.
+ *  \ingroup hardware_sleep
+ *
+ * One of the sleep_run_* functions must be called prior to this call
+ *
+ * \param delay_ms The duration to sleep for in milliseconds.
+ * \param callback Function to call on wakeup.
+ */
+void sleep_goto_sleep_for(uint32_t delay_ms, hardware_alarm_callback_t callback, int* alarm_id);
```
Comment (lurch) [inline src/rp2_common/pico_sleep/include/pico/sleep.h:88]: Missing `\param` description above for `alarm_id` ?

File: src/rp2_common/pico_sleep/sleep.c [hunk 6]
```diff
@@ -102,20 +143,77 @@ void sleep_run_from_dormant_source(dormant_source_t dormant_source) {
...
+#else
+#error Unknown processor
+#endif
+
+    int alarm_num = hardware_alarm_claim_unused(true);
+
+    // Return the timer alarm irq number (0..3) so it can be disabled and freed from the main program
+    *alarm_id = alarm_num;
+
+    uart_default_tx_wait_blocking();
```
Comment (lurch) [inline src/rp2_common/pico_sleep/sleep.c:204]: I wonder if it might be worth adding a comment explaining why this line is needed?

**Comment by pspeybro** [comment]
Is there any chance that this update fixes the problem with locking up the pico after repeatedly calling stdio_init_all(); or stdio via USB that stops working after waking up?
See this issue for some of the attempts to find a solution: https://github.com/raspberrypi/pico-extras/issues/41

File: src/rp2_common/pico_sleep/sleep.c (hunk 6 above)
Comment (peterharperuk) [reply src/rp2_common/pico_sleep/sleep.c:204]: I think that should be stdio_flush

File: src/rp2_common/pico_sleep/include/pico/sleep.h (hunk 1 above)
Comment (peterharperuk) [reply src/rp2_common/pico_sleep/include/pico/sleep.h:77]: actually I disagree with myself, lets leave it in

**Comment by peterharperuk** [comment]
I'm going to merge this...

**Review by kilograham — APPROVED** [review]

**Comment by Lecostarius** [comment]
Many thanks for looking into this!!
Can anyone confirm that the code is working for the Pico 2 (RP2350)? For me, the examples in pico_playground work nicely for the RP2040 (Pico), but not for the Pico 2 - it does not wake up. I can see however that the current consumption on the Pico 2 drops a lot when put to sleep, so the problem appears to be in the wake up part.

**Comment by Lecostarius** [comment]
OK, just saw your reply a moment ago. I will check it out and let you know. Thanks for the answer!

**Comment by Lecostarius** [comment]
I think I have used the latest update, but still no luck with pico2. I succeeded it switching it off completely using powman and it wakes up again, which is also solving my problem.
I should note that I was using the Arduino IDE for that which most likely is not a good idea.

**Comment by peterharperuk** [comment]
@Lecostarius There are a few sleep examples, sleep and dormant, are none of them waking up? They should wake up when an interrupt goes off - it's odd that this wouldn't happen. Usually they don't go to sleep. You haven't enabled stdio_usb or made any other changes?

**Comment by Lecostarius** [comment]
They work on Pico. On Pico2, no wakeup. I could get the powman based example in pico-examples to work on the Pico2 though - shutdown and wakeup both work. 

Probably the problem sits in front of the computer... if I find the time I will write an update here, but for now, I will go with the powman based shutdown.

**Pull request merged**